graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -395,13 +407,16 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
 
         if edge_traversal_is_optional:
             # Entering an optional block!
-            # Make sure there's a tag right before it for the optional Backtrack to jump back to.
-            # Otherwise, the traversal could rewind to an old tag and might ignore
-            # entire stretches of applied filtering.
-            if not isinstance(basic_blocks[-1], blocks.MarkLocation):
+            # Make sure there's a marked location right before it for the optional Backtrack
+            # to jump back to. Otherwise, the traversal could rewind to an old marked location
+            # and might ignore entire stretches of applied filtering.
+            if context['marked_location_stack'][-1].num_traverses != 0:
```

Comment:
This should technically be `> 0` no? I realize they end up doing the same thing, but it's nice to be explicit.